plugin/cql/src/main/java/org/opencds/cqf/ruler/cql/KnowledgeArtifactProcessor.java

```diff
@@ -574,6 +583,225 @@ private Optional<MetadataResource> checkIfReferenceInList(RelatedArtifact artifa
+			// entries list at any time
+			List<BundleEntryComponent> entries = packagedBundle.getEntry();
+			Integer bundleSize = entries.size();
+			if (count < bundleSize){
+				packagedBundle.setEntry(entries.subList(0, count));
+			} else {
+				// there are not enough entries in the bundle
+				// to page so we return all of them
+				// no change
+			}
+		}
+		return packagedBundle;
+	}
+	void recursivePackage(
+		MetadataResource resource,
+		Bundle bundle,
+		FhirDal fhirDal,
+		List<String> capability,
+		List<String> include,
+		List<CanonicalType> canonicalVersion,
+		List<CanonicalType> checkCanonicalVersion,
+		List<CanonicalType> forceCanonicalVersion
+		){
+		if (resource != null) {
+			KnowledgeArtifactAdapter<MetadataResource> adapter = new KnowledgeArtifactAdapter<MetadataResource>(resource);
+			findUnsupportedCapability(resource, capability);
+			processCanonicals(resource, canonicalVersion, checkCanonicalVersion, forceCanonicalVersion);
+			boolean entryExists = bundle.getEntry().stream().anyMatch(
+				e -> e.hasResource()
+					&& ((MetadataResource)e.getResource()).getUrl().equals(resource.getUrl())
+					&& ((MetadataResource)e.getResource()).getVersion().equals(resource.getVersion())
+			);
+			if (!entryExists) {
+				BundleEntryComponent entry = createEntry(resource);
+				// TODO: remove history from request.url
+				entry.getRequest().setMethod(HTTPVerb.POST);
+				entry.getRequest().setIfNoneExist("url="+resource.getUrl()+"&version="+resource.getVersion());
+				bundle.addEntry(entry);
+			}
+			List<RelatedArtifact> components = adapter.getComponents();
+			List<RelatedArtifact> dependencies = adapter.getDependencies();
+			Stream.concat(components.stream(), dependencies.stream())
+				.map(ra -> searchResourceByUrl(ra.getResource(), fhirDal))
+				.map(searchBundle -> searchBundle.getEntry().stream().findFirst().orElseGet(()-> new BundleEntryComponent()).getResource())
+				.forEach(component -> recursivePackage((MetadataResource)component, bundle, fhirDal, capability, include, 	canonicalVersion, checkCanonicalVersion, forceCanonicalVersion));
+		}
+	}
+	private Optional<String> findVersionInListMatchingResource(List<CanonicalType> list, MetadataResource resource){
+		return list.stream()
+					.filter((canonical) -> Canonicals.getUrl(canonical).equals(resource.getUrl()))
+					.map((canonical) -> Canonicals.getVersion(canonical))
+					.findAny();
+	}
+	private void findUnsupportedCapability(MetadataResource resource, List<String> capability) throws PreconditionFailedException{
+		if (capability != null) {
+			List<Extension> knowledgeCapabilityExtension = resource.getExtension().stream()
+			.filter(ext -> ext.getUrl().contains("cqf-knowledgeCapability"))
+			.collect(Collectors.toList());
+			if (knowledgeCapabilityExtension.isEmpty()) {
+				// consider resource unsupported
+				// if it's knowledgeCapability is
+				// undefined
+				throw new PreconditionFailedException("Artifact does not match capability");
```

Comment:
The phrasing is WIP still, but currently assuming if there is no KnowledgeCapability extension we should throw an error as well as for KnowledgeCapability where it doesn't match?

Reply:
This seems correct to me.